### ISSUE TITLE
Bug: False positives after update

### ISSUE THREAD

**Issue by plehnen** [typescript-eslint/typescript-eslint#7359]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

```typescript
After updating ... 

"@typescript-eslint/eslint-plugin": "^5.62.0",
"@typescript-eslint/parser": "^5.62.0",
```
...to a newer version, running eslint shows many `@typescript-eslint/no-unused-vars` errors, which aren't errors:

i.e. 

```
export const XY: InjectionKey<Ref<boolean | ((arg: string) => boolean)>> = Symbol('XY'); // 'arg' is defined but never used              @typescript-eslint/no-unused-vars

export type CustomFiltersType = Record<
  string,
  {
    label?: (value: any) => string; // 'value' is defined but never used  @typescript-eslint/no-unused-vars
  }

// in vue props:
    disabledDate: {
      type: Function as PropType<(date: string) => boolean>, // 'date' is defined but never used  @typescript-eslint/no-unused-vars
    },
>;
```
```


### ESLint Config

```javascript
module.exports = {
  root: true,

  env: {
    node: true,
    es2022: true,
  },
  extends: [
    'plugin:vue/vue3-recommended',
    'plugin:vue/vue3-strongly-recommended',
    'eslint:recommended',
    '@vue/prettier',
    'plugin:vuejs-accessibility/recommended',

    // https://github.com/vuejs/eslint-plugin-vue#bulb-rules
    'plugin:vue/recommended',
    'plugin:vue/vue3-essential',

    '@vue/typescript/recommended',
    '@vue/prettier',
    '@vue/eslint-config-typescript',
  ],

  parserOptions: {
    jsx: true,
    ecmaVersion: 2020,
    parser: '@typescript-eslint/parser',
    sourceType: 'module',
    tsconfigRootDir: __dirname,
  },

  plugins: ['import', 'promise', 'simple-import-sort', 'no-only-tests', 'vuejs-accessibility'],

  rules: {
    // Only allow debugger in development
    'no-debugger': process.env.PRE_COMMIT ? 'error' : 'off',
    // Only allow `console.log` in development
    'no-console': process.env.PRE_COMMIT ? ['error', { allow: ['warn', 'error'] }] : 'off',
    'import/no-relative-parent-imports': 'error',
    'no-restricted-imports': [
      'error',
      {
        patterns: [
          {
            group: ['./*/', '**../'],
            message: 'Relative imports are not allowed.',
          },
        ],
      },
    ],
    'vue/array-bracket-spacing': 'error',
    'vue/arrow-spacing': 'error',
    'vue/block-spacing': 'error',
    'vue/brace-style': 'error',
    'vue/camelcase': 'error',
    'vue/comma-dangle': ['error', 'always-multiline'],
    'vue/component-name-in-template-casing': 'error',
    'vue/dot-location': ['error', 'property'],
    'vue/eqeqeq': 'error',
    'vue/key-spacing': 'error',
    'vue/keyword-spacing': 'error',
    'vue/no-boolean-default': 'error',
    'vue/no-deprecated-scope-attribute': 'error',
    'vue/no-empty-pattern': 'error',
    'vue/object-curly-spacing': ['error', 'always'],
    'vue/padding-line-between-blocks': 'error',
    'vue/multi-word-component-names': 'off',
    'vue/v-on-event-hyphenation': [
      'error',
      'always',
      {
        autofix: true,
      },
    ],
    'vue/space-infix-ops': 'error',
    'vue/space-unary-ops': 'error',
    'vue/v-on-function-call': 'error',
    'vue/v-slot-style': 'error',
    'vue/valid-v-slot': 'error',
    'vue/prefer-import-from-vue': 'error',
    'vue/prefer-true-attribute-shorthand': 'error', // NOTE: properties with multiple types are not defaulting to true! in this case you can explicitly use :value="!!true"
    'vue/attribute-hyphenation': 'error',
    'vue/html-self-closing': [
      'error',
      {
        html: {
          void: 'any',
          normal: 'any',
        },
        svg: 'any',
      },
    ],
    'vue/no-empty-component-block': 'error',
    'vue/no-use-computed-property-like-method': 'error',
    'vue/no-useless-mustaches': 'error',
    'vue/no-useless-v-bind': 'error',
    'vue/prefer-separate-static-class': 'error',
    'vue/v-for-delimiter-style': 'error',
    eqeqeq: 'error',
    'no-useless-return': 'error',
    'no-useless-rename': 'error',
    'no-return-await': 'error',
    'max-classes-per-file': 'off',
    'no-empty-function': 'off',
    'import/prefer-default-export': 'off',
    'no-use-before-define': 'off',
    '@typescript-eslint/no-unused-vars': ['warn'],
    '@typescript-eslint/no-explicit-any': ['off'],
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/ban-ts-comment': [
      'error',
      {
        'ts-ignore': 'allow-with-description',
        minimumDescriptionLength: 5,
      },
    ],
    semi: [2, 'always'],
    'vue/no-v-html': 'off',
    'no-only-tests/no-only-tests': process.env.PRE_COMMIT ? 'error' : 'off',
    'vuejs-accessibility/label-has-for': [
      2,
      {
        controlComponents: ['ElInput'],
        required: {
          some: ['nesting', 'id'],
        },
      },
    ],
    'vue/no-multiple-template-root': 0,
    'vue/no-v-model-argument': 0,
    'vue/no-v-for-template-key': 0,
    'vue/script-setup-uses-vars': 0,
    'prefer-arrow-callback': 'error',
    // 'func-style': 'error', // currently not recommended, because some typescript functions (see assertions.ts or SimpleTable.vue) needs to be written as plain functions
    'func-names': 'error',
    'simple-import-sort/imports': [
      'error',
      {
        groups: [
          [
            '^vue', // vue*
            '@route.*', // @route*
            '^@?\\w', // 3rd party libraries
            '^', // all remaining imports sorted alphabetically
            '^@/utils(/.*|$)', // utils
            '^@/modules(/.*|$)', // modules
            '^.+\\.s?css$', // css
            '^\\.', // relative import
            '^\\u0000', // type imports...
            '^@?\\w.*\\u0000$',
            '^[^.].*\\u0000$',
            '^\\..*\\u0000$',
          ],
        ],
      },
    ],
    'simple-import-sort/exports': 'error',
    'import/first': 'error',
    'import/newline-after-import': 'error',
    'import/no-duplicates': 'error',
  },

  globals: {
    require: true,
    process: true,
    module: true,
    __dirname: true,
  },

  overrides: [
    {
      files: ['tests/unit/**/*', 'tests/e2e/**/*', '**/__tests__/*.{j,t}s?(x)', '**/*.{spec,unit,cy}.{j,t}s?(x)'],
      rules: {
        'no-restricted-imports': 'off',
      },
    },
    {
      files: ['src/**/*', 'tests/unit/**/*', 'tests/e2e/**/*'],
      env: {
        browser: true,
      },
    },
    {
      files: ['**/__tests__/*.{j,t}s?(x)', '**/*.{spec,unit}.{j,t}s?(x)'],
      globals: {
        mount: false,
        shallowMount: false,
        shallowMountView: false,
        createComponentMocks: false,
        createModuleStore: false,
      },
    },
    {
      files: ['**/*.{vue,ts,tsx}'],
      extends: ['@vue/typescript/recommended', '@vue/eslint-config-typescript'],
      rules: {
        '@typescript-eslint/no-unused-vars': ['warn'],
        '@typescript-eslint/no-explicit-any': ['off'],
        '@typescript-eslint/explicit-module-boundary-types': 'off',
      },
    },
    {
      files: ['**/__tests__/*.{j,t}s?(x)', '**/tests/unit/**/*.spec.{j,t}s?(x)'],
    },
  ],
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "strict": true,
    "outDir": "./dist",
    "noImplicitAny": false,
    "strictNullChecks": true,
    "jsx": "preserve",
    "importHelpers": true,
    "moduleResolution": "node",
    "isolatedModules": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "experimentalDecorators": true,
    "allowJs": true,
    "sourceMap": true,
    "resolveJsonModule": true,
    "baseUrl": ".",
    "removeComments": true,
    "typeRoots": ["node_modules/@types", "src/@types"],
    "types": ["vite/client", "cypress", "element-plus/global", "@intlify/unplugin-vue-i18n/messages"],
    "paths": {
      "@/*": ["src/*", "src/*.vue", "src/*/index.vue", "src/*/index.js", "src/*/index.ts"],
      "@route-types": ["src/router/route-types.ts"],
      "@routes": ["src/router/routes.ts"]
    },
    "lib": ["esnext", "dom", "dom.iterable", "scripthost"]
  },
  "include": ["src/**/*.ts", "src/**/*.tsx", "src/**/*.vue", "tests/**/*.ts", "tests/**/*.tsx", "components.d.ts", "auto-imports.d.ts"],
  "exclude": ["node_modules", "dist", "tests/**/*", "src/**/*.unit.ts", "src/**/*.spec.ts", "src/**/*.test.ts"]
}
```

### Expected Result

No false positive errors after updating.

### Actual Result

Many errors which aren't errors.

### Versions

"dependencies": {
    "@element-plus/icons-vue": "^2.1.0",
    "@fortawesome/fontawesome-common-types": "6.4.0",
    "@fortawesome/fontawesome-svg-core": "6.4.0",
    "@fortawesome/free-regular-svg-icons": "6.4.0",
    "@fortawesome/free-solid-svg-icons": "6.4.0",
    "@fortawesome/vue-fontawesome": "3.0.3",
    "@intlify/unplugin-vue-i18n": "^0.12.2",
    "@microsoft/applicationinsights-web": "^3.0.2",
    "@popperjs/core": "^2.11.8",
    "@vueuse/core": "^10.2.1",
    "axios": "1.4.0",
    "chart.js": "^3.9.1",
    "chartjs-adapter-dayjs-3": "^1.2.3",
    "core-js": "^3.31.1",
    "css-pro-layout": "^1.1.0",
    "dayjs": "1.11.9",
    "dompurify": "^3.0.5",
    "element-plus": "2.3.5",
    "lodash-es": "4.17.21",
    "marked": "^5.1.1",
    "modern-normalize": "^2.0.0",
    "pinia": "^2.1.4",
    "tslib": "^2.6.0",
    "util": "^0.12.5",
    "vue": "^3.3.4",
    "vue-chart-3": "^3.1.8",
    "vue-focus-lock": "2.0.5",
    "vue-global-events": "3.0.0",
    "vue-i18n": "9.2.2",
    "vue-router": "4.2.4",
    "zod": "^3.21.4"
  },
  "devDependencies": {
    "@cypress/code-coverage": "^3.11.0",
    "@istanbuljs/nyc-config-typescript": "^1.0.2",
    "@playwright/test": "^1.36.2",
    "@total-typescript/ts-reset": "^0.4.2",
    "@types/dompurify": "^3.0.2",
    "@types/lodash-es": "^4.17.8",
    "@types/marked": "^5.0.1",
    "@typescript-eslint/eslint-plugin": "^6.2.0",
    "@typescript-eslint/parser": "^6.2.0",
    "@vitejs/plugin-vue": "^4.2.3",
    "@vitest/coverage-v8": "^0.33.0",
    "@vue/compiler-dom": "^3.3.4",
    "@vue/eslint-config-prettier": "^8.0.0",
    "@vue/eslint-config-typescript": "^11.0.3",
    "@vue/server-renderer": "^3.3.4",
    "@vue/test-utils": "^2.4.1",
    "change-case": "^4.1.2",
    "chokidar": "^3.5.3",
    "concurrently": "^8.2.0",
    "cross-env": "7.0.3",
    "csstype": "^3.1.2",
    "cypress": "^12.17.2",
    "cypress-intercept-formdata": "^0.5.2",
    "cypress-vite": "^1.4.2",
    "eslint": "8.45.0",
    "eslint-plugin-cypress": "^2.13.3",
    "eslint-plugin-import": "^2.27.5",
    "eslint-plugin-no-only-tests": "3.1.0",
    "eslint-plugin-prettier": "^5.0.0",
    "eslint-plugin-promise": "6.1.1",
    "eslint-plugin-simple-import-sort": "^10.0.0",
    "eslint-plugin-vue": "9.15.1",
    "eslint-plugin-vuejs-accessibility": "2.1.0",
    "husky": "^8.0.3",
    "istanbul-lib-coverage": "^3.2.0",
    "jsdom": "^22.1.0",
    "json-server": "^0.17.3",
    "lint-staged": "^13.2.3",
    "nyc": "^15.1.0",
    "prettier": "^3.0.0",
    "sass": "^1.64.1",
    "ts-essentials": "^9.3.2",
    "typescript": "4.9.5",
    "unplugin-auto-import": "^0.16.6",
    "unplugin-vue-components": "^0.25.1",
    "vite": "^4.4.6",
    "vite-bundle-visualizer": "^0.8.1",
    "vite-plugin-istanbul": "^5.0.0",
    "vitest": "^0.33.0",
    "vue-tsc": "^1.8.6",
    "wait-on": "^7.0.1"
  },

**Comment by bradzacher** [comment]
> Playground Link
> No response

Please use the "🐛 Report a Complex Bug With a Reproduction Repository" template and fill it out in full.
An issue with some config is impossible for us to help you with.
